### PR TITLE
Feature runtime parametrics

### PR DESCRIPTION
**Pull request by kantholz44mm**
implements a tokenizer for math expressions, handling single-letter variables ('x', 't', etc.), function names ('cos', 'log', etc.), basic arithmetic ('+', '-', '*', etc.) and parentheses. Implements the shunting yard algorithm to convert infix to postfix, as well as a stack based executor for such token lists.

### CODE REVIEW

**Comment by FabianRolfMatthiasNoll** [comment]
Just a thought: I dont know if we should add error handling? Does it make sense to handle the error and then return it so we can display it? Like badInput or emptyInput?

File: src/MathExpression/ShuntingYard.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+
+import MathExpression.Token.Type;
+
+public class ShuntingYard {
+
+    private static void processToken(Token token, List<Token> postfix, Stack<Token> shunt) {
+        switch(token.type) {
```
Comment (FabianRolfMatthiasNoll) [inline src/MathExpression/ShuntingYard.java:12]: It is not needed but i find the following style to write such switch cases in this scenario a little bit more readable:
```suggestion
       switch (token.type) {
            case Number, Parameter -> {
                postfix.add(token);
            }
```
and so forth... But thats for you to decide whichever you like best.

File: src/FunctionGrapher.java [hunk 2]
```diff
@@ -1,4 +1,5 @@
-import javax.swing.*;
+import javax.swing.JFrame;
+import javax.swing.SwingUtilities;
 
 public class FunctionGrapher {
```
Comment (FabianRolfMatthiasNoll) [inline src/FunctionGrapher.java:4]: A expression like "cos(cos(t) * -2) * 2 ^ 2" results in errors. Negative numbers are currently not handled

File: src/MathExpression/Token.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        Exponentiation,
+        Division,
+        Multiplication,
+        Addition,
+        Subtraction,
+        Function,
+        Parenthesis_Left,
+        Parenthesis_Right;
+
+        public boolean isOperator() {
```
Comment (FabianRolfMatthiasNoll) [inline src/MathExpression/Token.java:17]: same here could be written as 4-liner. Is it needed? Not really but i like it 👍 
```suggestion
        public boolean isOperator() {
            return switch (this) {
                case Exponentiation, Division, Multiplication, Addition, Subtraction, Function -> true;
                default -> false;
            };
        }
```

**Review by FabianRolfMatthiasNoll — COMMENTED** [review]
Looks really good. Only real point is the handling of negative values

File: src/MathExpression/Token.java (hunk 3 above)
Comment (kantholz44mm) [reply src/MathExpression/Token.java:17]: Didn't know about this notation, here it makes for a lot cleaner code. Done

File: src/FunctionGrapher.java (hunk 2 above)
Comment (kantholz44mm) [reply src/FunctionGrapher.java:4]: support for unary '-' added.

File: src/MathExpression/ShuntingYard.java (hunk 1 above)
Comment (kantholz44mm) [reply src/MathExpression/ShuntingYard.java:12]: In the case of the large switch in ShuntingYard, it is more readable the way it is, I think

Comment (FabianRolfMatthiasNoll) [reply src/MathExpression/ShuntingYard.java:12]: Thats ok ^^

**Review by FabianRolfMatthiasNoll — APPROVED** [review]
Looking good

**Pull request merged**